### PR TITLE
Add Azure Resource Abbreviations File

### PR DESCRIPTION
**Pull request by jongio**
Right now we have all our prefixes hardcoded. We want them to be consistent across all our base templates. If we don't have them shared they are likely to diverge. I would like to keep them all in a single file and reference them in all the templates using the bicep load json feature.

This is dependent on repoman being able to do arbitrary find/replace operations to fix up the paths.

While adding prefixes, we discovered that we deviated from the CAF convention by using -rg suffix instead of rg- prefix for resource groups.

We want to move to the rg- prefix, but also don't want to break existing templates that use -rg suffix.

Here's the approach that has been implemented in this PR.

## Defaults
1. Do not have a hard-coded default in go code
1. Default to "rg-envname" in bicep

## Resource Group discovery

1. If RG is in env var or azure.yaml then use it
1. If not, then query graph to get the first RG tagged with current azd-env-name.
    1. If found, use it
    1. If more than one found, show error message that multiples were found and they need to specify the RG name in azure.yaml.
    1. If not found, then try rg-envname, if not found try envname-rg (for backwards compat)
    1. If no possible RG found, then show error message explaining what they need to do.

### CODE REVIEW

**Comment by ellismg** [comment]
I think I'm onboard with this approach - the one sort of downside here is it's another piece of complexity in these templates - and yet at the same time maybe this is what best practices would look like for a larger organization.

Perhaps at some point we could try to be clever and make the emitted JSON file with prefixes just contain what it needs for the template we build out, might make things more approachable.

**Comment by jongio** [comment]
I thought about that… But we want people to extend the templates and I would like all the prefixes to be available when they do so

File: templates/todo/projects/nodejs-mongo/infra/applicationinsights.bicep [hunk 1]
```diff
@@ -3,8 +3,10 @@ param location string
 param tags object
 param workspaceId string
 
+var abbrs = loadJsonContent('../../../../common/infra/abbreviations.json')
```
Comment (wbreza) [inline templates/todo/projects/nodejs-mongo/infra/applicationinsights.bicep:6]: Just need to add the following to your `repo.yaml`

```yaml
  rewrite:
    patterns:
      - "**/*.@(yml|yaml)"
      - "**/*.bicep"
```

See https://github.com/Azure/azure-dev/pull/162/files#diff-b0ec1cd3c68dbde790266bd1136ee50c4e1a8fd06aad87bf7e79daf97efba12e for an example.

Comment (jongio) [reply templates/todo/projects/nodejs-mongo/infra/applicationinsights.bicep]: Got it.  Thanks!

**Comment by jongio** [comment]
Template tests have run successfully

File: cli/azd/pkg/azureutil/resource_group.go [hunk 2]
```diff
@@ -38,3 +41,79 @@ func GetResourceGroupsForDeployment(ctx context.Context, azCli tools.AzCli, subs
...
+	query := fmt.Sprintf(`resourceContainers 
+		| where type == "microsoft.resources/subscriptions/resourcegroups" 
+		| where tags['azd-env-name'] == '%s' 
+		| project id, name, type, tags, location`,
+		strings.ToLower(env.GetEnvName()))
+
+	queryResult, err := azCli.GraphQuery(ctx, query, []string{env.GetSubscriptionId()})
+
+	if err != nil {
+		return nil, fmt.Errorf("executing graph query: %s:%w", query, err)
```
Comment (weikanglim) [inline cli/azd/pkg/azureutil/resource_group.go:57]:
```suggestion
		return nil, fmt.Errorf("executing graph query: %s: %w", query, err)
```

File: cli/azd/pkg/azureutil/resource_group.go [hunk 3]
```diff
@@ -38,3 +41,79 @@ func GetResourceGroupsForDeployment(ctx context.Context, azCli tools.AzCli, subs
...
+	query := fmt.Sprintf(`resourceContainers 
+		| where type == "microsoft.resources/subscriptions/resourcegroups" 
+		| where name in('rg-%[1]s','%[1]s-rg')
+		| project id, name, type, tags, location`,
+		strings.ToLower(env.GetEnvName()))
+
+	queryResult, err := azCli.GraphQuery(ctx, query, []string{env.GetSubscriptionId()})
+
+	if err != nil {
+		return nil, fmt.Errorf("executing graph query: %s:%w", query, err)
```
Comment (weikanglim) [inline cli/azd/pkg/azureutil/resource_group.go:77]:
```suggestion
		return nil, fmt.Errorf("executing graph query: %s: %w", query, err)
```

**Review by weikanglim — APPROVED** [review]
Having a resource abbreviation file still makes sense even in the future as we make resource naming generation smarter / more customizable as basic developer building blocks.. LGTM.

File: generators/repo/src/commands/generate.ts [hunk 4]
```diff
@@ -419,4 +419,4 @@ export class GenerateCommand implements RepomanCommand {
             await fs.writeFile(destFilePath, contents);
         }
     }
-}
+}
```
[the remark this answers is not in the document]
Comment (jongio) [reply generators/repo/src/commands/generate.ts:422]: I think it is going to be more work to revert that line than it is worth it and we have inconsistency on .ts files that we should bulk fix.

**Review by vhvb1989 — APPROVED** [review]
a few comments, but LGTM

File: cli/azd/CONTRIBUTING.md [hunk 5]
```diff
@@ -79,7 +79,9 @@ Run `golangci-lint run ./...`
 
 ## Spell checking
 
-Install [cspell](https://cspell.org/) and then run `cspell lint "**/*.go" --relative --config ./.vscode/cspell.yaml`
+1. Install [cspell](https://cspell.org/)
+2. CD to /cli/azd
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/CONTRIBUTING.md:83]: I ran into the issue of not being in the right folder. So, I/d like to just keep it as-is

File: cli/azd/pkg/azureutil/resource_group.go [hunk 6]
```diff
@@ -38,3 +41,79 @@ func GetResourceGroupsForDeployment(ctx context.Context, azCli tools.AzCli, subs
 
 	return keys, nil
 }
+
+// GetResourceGroupsForEnvironment gets all resources groups for a given environment
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/pkg/azureutil/resource_group.go:45]: We don't prevent it, but our templates only have one.

File: cli/azd/pkg/environment/names.go [hunk 7]
```diff
@@ -3,14 +3,10 @@
...
 package environment
 
-// DefaultResourceGroupName returns the resource group name used by default for projects.
-// Returns environment variable value by default
-// otherwise uses convention, `{AZURE_ENV_NAME}-rg`.
-func DefaultResourceGroupName(env *Environment) string {
+func GetResourceGroupNameFromEnvVar(env *Environment) string {
 	resourceGroupName, ok := env.Values[ResourceGroupEnvVarName]
 	if ok {
 		return resourceGroupName
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/pkg/environment/names.go:9]: Created an issue to track that validation

File: cli/azd/CONTRIBUTING.md [hunk 8]
```diff
@@ -79,7 +79,9 @@ Run `golangci-lint run ./...`
...
+3. Run `cspell lint "**/*.go" --relative --config ./.vscode/cspell.yaml`
```
Comment (karolz-ms) [inline cli/azd/CONTRIBUTING.md:84]: Not necessarily for this PR, but we should also mention the cspell extension for VS Code, which allows you to track spelling issues "on the fly", without having to run cspell manually.

File: cli/azd/pkg/azureutil/resource_group.go [hunk 9]
```diff
@@ -38,3 +41,80 @@ func GetResourceGroupsForDeployment(ctx context.Context, azCli tools.AzCli, subs
...
+func GetResourceGroupsForEnvironment(ctx context.Context, env *environment.Environment) ([]tools.AzCliResource, error) {
```
Comment (wbreza) [inline cli/azd/pkg/azureutil/resource_group.go:46]: Can you move these into `pkg/infra/azure_resource_manager.go` ?

File: cli/azd/test/samples/funcapp/infra/resources.bicep [hunk 10]
```diff
@@ -1,11 +1,12 @@
...
+param tags object
 
-resource function 'Microsoft.Web/sites@2021-03-01' = {
-  name: '${name}func'
+resource function 'Microsoft.Web/sites@2022-03-01' = {
+  name: 'func-${resourceToken}'
   location: location
   kind: 'functionapp,linux'
-
+  tags: union(tags, { 'azd-service-name': 'func' })
```
Comment (wbreza) [inline cli/azd/test/samples/funcapp/infra/resources.bicep:9]: Is `func` going to work here? This usually needs to be the actual azure resource name and `func` sounds too vague to be unique in azure.

File: cli/azd/test/samples/resourcegroups/infra/main.parameters.json [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    "resourceGroupName": {
+      "value": "${TEST_RESOURCE_GROUP_NAME}"
+    },
+    "includeEnvNameTag": {
+      "value": "${TEST_INCLUDE_ENV_NAME_TAG}"
+    }
+    ,
+    "createMultipleResourceGroups": {
+      "value": "${TEST_CREATE_MULTIPLE_RESOURCE_GROUPS}"
+    }
```
Comment (wbreza) [inline cli/azd/test/samples/resourcegroups/infra/main.parameters.json:20]: I'm assuming these are added here just to support better test cases and not something we would propose in our templates?

**Review by wbreza — APPROVED** [review]
Added a couple minor comments but otherwise looks great!

**Review by karolz-ms — APPROVED** [review]
Thanks for taking care of this change, nice improvement!

File: cli/azd/CONTRIBUTING.md (hunk 8 above)
Comment (vhvb1989) [reply cli/azd/CONTRIBUTING.md:84]: It is part of the recommended extensions: https://github.com/Azure/azure-dev/blob/main/cli/azd/.vscode/extensions.json#L6

File: cli/azd/pkg/azureutil/resource_group.go (hunk 9 above)
Comment (jongio) [reply cli/azd/pkg/azureutil/resource_group.go:46]: Which methods?  It wasn't clear to me which ones should go in what file.

File: cli/azd/test/samples/funcapp/infra/resources.bicep (hunk 10 above)
Comment (jongio) [reply cli/azd/test/samples/funcapp/infra/resources.bicep:9]: Func maps to the key in azure.yaml

File: cli/azd/test/samples/resourcegroups/infra/main.parameters.json (hunk 11 above)
Comment (jongio) [reply cli/azd/test/samples/resourcegroups/infra/main.parameters.json:20]: Correct, test helpers only.

File: cli/azd/CONTRIBUTING.md (hunk 8 above)
Comment (jongio) [reply cli/azd/CONTRIBUTING.md:84]: I created an issue to track this.

File: cli/azd/pkg/azureutil/resource_group.go (hunk 9 above)
Comment (jongio) [reply cli/azd/pkg/azureutil/resource_group.go:46]: Discussed offline.  Wallace will move them in his upcoming PR.

File: cli/azd/test/samples/funcapp/infra/resources.bicep (hunk 10 above)
Comment (wbreza) [reply cli/azd/test/samples/funcapp/infra/resources.bicep:9]: Yup - you're right

**Pull request merged**